[Fiber] Force rerender a failed tree

acdlite opened the pull request:
This lets us preserve the invariant that if a tree contains pending work, its root is scheduled. Previously, a tree with an uncaught error broke this assumption.

Now we can exit early in `scheduleUpdate` once we reach a node whose priority matches.

Closes https://github.com/facebook/react/issues/8222

sebmarkbage commented:
Curious to hear what @gaearon thinks about this. I think it is possible to unmount it from React's perspective but avoid removing the child at the root if we want to leave it in.

gaearon commented:
I think the least surprising behavior is:

* All unmount lifecycles fire (safely, i.e. errors in them are ignored)
* `isMounted()` for components returns `false` and `setState()` is a no-op
* All event handlers are removed, but the DOM tree stays at the same spot, it's just "non-interactive" now
* Next `unmountComponentAtNode()` removes the DOM tree
* Next `render()` also removes the DOM tree, and mounts a new one

acdlite commented:
Even if we leave the DOM in, should React attempt to re-use the previous, failed work or start from scratch?

acdlite commented:
I assume we should do the same thing for caught error boundaries? "Unmount" the tree, without actually removing the DOM nodes, and mount a new one the next time the boundary is re-rendered.

gaearon commented:
Would the difference be observable if boundaries are processed synchronously?

acdlite commented:
Yes because the entire failed tree below the boundary is unmounted regardless, whereas in the current implementation parts of the failed tree may be reused.

sebmarkbage commented:
I just want to frame the discussion a bit:

The worst case scenario isn't that the user get frustrated that the UI is non-interactive or confusing. That's the best case. We have to weight that against the worst case.

The worst case, would likely be that you get into an inconsistent state so that internal values are not represented on the screen and then an action gets taken based on that. For example, a user spends the wrong amount, pays for the wrong thing, or worse, "Like" your ex's photo instead of your current significant other.

I think it is very easy to get in to an inconsistent state for a subtree that has errored. Given that life-cycles can mutate instances etc. Recovering from that may work in 99% of cases but in the 1% it doesn't, it can be devastating.

So, my preference would be to force an unmount even in the error boundary case. So that the view gets a forced rerender.

However, in the forced rerender case, when do you remove the child node? You don't want to remove it in case the error boundary then fails, because then it is gone. I guess you don't, and just let the regular reconciliation remove it if it succeeds.

acdlite commented:
> However, in the forced rerender case, when do you remove the child node? You don't want to remove it in case the error boundary then fails, because then it is gone. I guess you don't, and just let the regular reconciliation remove it if it succeeds.

That makes sense to me.

I'll give this a shot.

gaearon commented:
> So, my preference would be to force an unmount even in the error boundary case. So that the view gets a forced rerender.

👍

acdlite commented:
@gaearon Ready for review!

Now that boundary's are force remounted, I had to tweak some of the unit tests in `ReactErrorBoundaries-test.js`. I believe the changes I made correctly match the new semantics but they might deserve some extra scrutiny.

gaearon commented on src/renderers/shared/stack/reconciler/__tests__/ReactErrorBoundaries-test.js:
```diff
@@ -1047,8 +1047,10 @@ describe('ReactErrorBoundaries', () => {
-      'ErrorBoundary unstable_handleError',
       ...(ReactDOMFeatureFlags.useFiber ? [
+        // Fiber unmounts an error boundary's children
+        'Normal componentWillUnmount',
```
There is another `Normal componentWillUnmount` below right before the array ends.
Does this means we now call it twice? Doesn't seem right.

gaearon commented on src/renderers/shared/stack/reconciler/__tests__/ReactErrorBoundaries-test.js:
```diff
@@ -1105,8 +1108,10 @@ describe('ReactErrorBoundaries', () => {
-      'ErrorBoundary unstable_handleError',
       ...(ReactDOMFeatureFlags.useFiber ? [
+        // Fiber unmounts an error boundary's children
+        'Normal componentWillUnmount',
```
Same here. There's another `Normal componentWillUnmount` later.

gaearon commented on src/renderers/shared/stack/reconciler/__tests__/ReactErrorBoundaries-test.js:
```diff
@@ -1158,8 +1164,11 @@ describe('ReactErrorBoundaries', () => {
       ...(ReactDOMFeatureFlags.useFiber ? [
+        // Fiber unmounts an error boundary's children
+        'Normal componentWillUnmount',
+        'BrokenComponentWillReceiveProps componentWillUnmount',
```
This one is also duplicated: see `BrokenComponentWillReceiveProps componentWillUnmount` later in the array.

gaearon commented on src/renderers/shared/fiber/ReactFiberCommitWork.js:
```diff
@@ -193,11 +195,14 @@ module.exports = function<T, P, I, TI, C>(
         commitNestedUnmounts(node);
```
Does this call (or any `commitUnmount()` calls it makes) also need to be guarded by `!ignoreTop`?
I don't fully understand what's happening here but it offers another path to `commitUnmount()`.

gaearon commented on src/renderers/shared/fiber/ReactFiberCommitWork.js:
```diff
@@ -193,11 +195,14 @@ module.exports = function<T, P, I, TI, C>(
         // node from the tree.
-        if (parent) {
+        if (parent && !ignoreHostNode) {
           removeChild(parent, node.stateNode);
```
Please bear with me as I don't fully understand this method yet.
Can `removeChild()` below run more than once?
If it can, `ignoreHostNode` should be named `ignoreHostNodes`?

gaearon commented on src/renderers/shared/fiber/ReactFiberCommitWork.js:
```diff
@@ -193,11 +195,14 @@ module.exports = function<T, P, I, TI, C>(
       } else {
-        commitUnmount(node);
+        // If ignoreTop is true, don't unmount the top fiber
+        if (!ignoreTop || node !== current) {
```
Similar question here. Why are we sure this is a "top" node? It isn't quite clear to me.
Is there only one "top" node, or can this block run more than twice in a loop?
Is there a better name for `ignoreTop`? I don't understand which fiber exactly would be called a "top" node.

gaearon replied:
Maybe it would be easier to understand if the comment emphasized when this happens (e.g. "If node and current are different, it means X. When `ignoreTop` is set, it means Y. Therefore we do Z to implement this [scenario described from user's point of view].")

gaearon commented on src/renderers/shared/fiber/ReactFiberCommitWork.js:
```diff
@@ -236,6 +241,34 @@ module.exports = function<T, P, I, TI, C>(
+    // the tree errored, we won't actually remove the host nodes; we'll leave
+    // them there until the tree is rerendered.
+    unmountHostComponents(null, current, true, false);
+    current.child = null;
```
Do we need to clear `progressedChild`?

gaearon commented on src/renderers/shared/fiber/ReactFiberCommitWork.js:
```diff
@@ -236,6 +241,34 @@ module.exports = function<T, P, I, TI, C>(
+      unmountHostComponents(parent, current, true, true);
+    }
+
+    boundary.child = null;
```
Same question about `progressedChild` here.

gaearon commented on src/renderers/shared/fiber/ReactFiberScheduler.js:
```diff
@@ -558,22 +564,22 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
+            root = nextScheduledRoot;
+            while (root) {
+              forceUnmountFailedRoot(root);
+              root.current.pendingWorkPriority = NoWork;
```
Do we need to update `root.current.alternate` here?

gaearon commented on src/renderers/shared/fiber/ReactFiberScheduler.js:
```diff
@@ -753,27 +753,31 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
       priorityLevel = TaskPriority;
     }
```
What does this commit fix? Is there a test that was failing before? Is this a performance optimization?

acdlite replied:
Perf optimization, unobservable

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberCommitWork.js:
```diff
@@ -184,7 +185,7 @@ module.exports = function<T, P, I, TI, C>(
   }
 
-  function unmountHostComponents(parent, current): void {
+  function unmountHostComponents(parent, current, ignoreHostNode, ignoreTop): void {
```
I think we can find a way to not use these flags.

A normal unmount won't schedule a removeChild anywhere but the top root where it was deleted. All we have to do is special case the root and then the rest of the traversal can be exactly the same.

Similarly, ignoreTop wouldn't be needed if instead we did the root work outside this function instead of inside it. That way we don't have to branch inside the loop.

sebmarkbage commented:
I worry about us making mutations directly to the tree. It makes it a bit difficult to reason about what will happen and what needs to be done to get the tree into a consistent state.

I think we should strive to reuse as much of the normal reconciliation logic as possible when we deal with errors. So that we know that everything gets processed in the same way.

I wonder if we can just do a normal reconciliation but force ReactChildFiber to not reuse the previous  node. That way all the regular processing kicks in just as if children's key differ.

acdlite commented:
I found a few other bugs last night after I pushed this... Going to try @sebmarkbage's suggested approach.

acdlite commented:
If we decide to go forward with this overall approach I'll add error boundary tests for each of the different types of children (corresponding to `reconcileChildrenArray`, `reconcileSingleElement`, etc.).

acdlite commented:
Btw I'm aware there are some mistakes in my edits to `ReactChildFiber`. Fixing it up now (learning more about how it works as I go) but wanted to push this up to get overall feedback on the approach.

acdlite commented:
I pushed some more changes. Implemented the failed root case using a flag on the FiberRoot. Added the concept of a root "error phase."

- The `SoftDeletion` phase happens when the error is first caught, and indicates that the tree should be unmounted without removing any host nodes.
- The `HardDeletion` phase indicates to `ReactChildFiber` that the next time the container is rendered into, the old nodes should be force removed from the DOM. This uses the normal reconciliation logic, as with the error boundary case.
- The default phase is `NoError`.

These phases are only used for the failed root case. They're not necessary for the error boundary case because the nodes are immediately replaced by a new tree.

I'll wait for feedback before I add any more test coverage. Still not sure how to avoid a flag on the class instance (_isFailedErrorBoundary).

gaearon commented:
Is `activeErrorBoundaries.has()` an adequate check for the case you implement with a flag? You could expose that to CommitWork too. If not maybe you could have a separate set for boundaries in that phase?

acdlite commented:
@gaearon Yes I could thread a parameter through based on `activeErrorBoundaries`. It'd be a bit cumbersome because I need in inside `ReactChildFiber`, which is connected to the scheduler via `ReactFiberBeginWork`. I held off doing that at first to see if there was some way we could infer a failed error boundary without an extra field. Don't think there is, though.

acdlite commented on src/renderers/shared/fiber/ReactFiberBeginWork.js:
```diff
@@ -59,370 +60,387 @@ module.exports = function<T, P, I, TI, C>(
-      // continue from this point.
-      current.progressedChild = workInProgress.progressedChild;
-      current.progressedPriority = workInProgress.progressedPriority;
+  function BeginWork(forceReplace : boolean) {
```
The changes in this file are minimal; there's a lot of noise because I wrapped everything in this function, which increased the indentation.

The only real change is the addition of a `forceReplace` parameter.

acdlite commented:
Latest commit removes the temporary `_isFailedErrorBoundary` hack. Good for another review.

@sebmarkbage I wrapped `ReactFiberBeginWork` in a wrapper function so that I could branch on whether the work in progress is a failed node, following the same pattern as `ReactChildFiber`. This seemed like a reasonable solution to me but perhaps I'm overlooking something.

joh1234 commented:
Update README
https://goo.gl/JBljdR

joh1234 commented:
Update README
https://goo.gl/JBljdR

or 
http://bit.ly/2dK6Ojk

gaearon commented:
This was superseded by #8304.
